homeassistant/components/azuredns/__init__.py

```diff
@@ -0,0 +1,176 @@
+    clientsecret = config[DOMAIN][CONF_CLIENTSECRET]
+    subscriptionid = config[DOMAIN][CONF_SUBSCRIPTIONID]
+    resourcegroupname = config[DOMAIN][CONF_RESOURCEGROUPNAME]
+    timeout = config[DOMAIN][CONF_TIMEOUT]
+    ttl = config[DOMAIN][CONF_TTL]
+    authority_url = AUTHORITYHOSTURL + '/' + config[DOMAIN][CONF_TENANT]
+    api_url = ('https://management.azure.com/subscriptions/' +
+               subscriptionid +
+               '/resourceGroups/' +
+               resourcegroupname +
+               '/providers/Microsoft.Network/dnsZones/' +
+               domain +
+               '/A/' +
+               host +
+               '?api-version=2018-05-01')
+
+    async_get_clientsession(hass)
+
+    result = await _update_azuredns(resource, tenant,
+                                    clientid, clientsecret,
+                                    authority_url, api_url, timeout, ttl)
+
+    if not result:
+        _LOGGER.error("Failed to update Azure DNS record.")
+        return False
+
+    async def update_domain_interval():
+        """Update the Azure DNS entry."""
+        await _update_azuredns(resource, tenant,
+                               clientid, clientsecret, authority_url,
+                               api_url, timeout, ttl)
+
+    async_track_time_interval(hass, update_domain_interval, INTERVAL)
+    return result
+
+
+async def _update_azuredns(resource, tenant,
+                           clientid, clientsecret, authority_url,
+                           api_url, timeout, ttl):
+    """Update the Azure DNS Record with the external IP address."""
+    import adal
+    import aiohttp
+
+    params = {
+        'resource': resource,
+        'tenant': tenant,
+        'clientid': clientid,
+        'clientsecret': clientsecret,
+        'api_url': api_url,
+    }
+
+    # Acquire the Azure AD token.
+    context = adal.AuthenticationContext(
+        authority_url, validate_authority=['tenant'] != 'adfs',
+    )
+
+    token = context.acquire_token_with_client_credentials(
+        params['resource'],
+        params['clientid'],
+        params['clientsecret'])
+
+    access_token = token.get('accessToken')
+
+    if not access_token:
+        _LOGGER.error("Failed to acquire Azure AD Access Token.")
+    else:
+        _LOGGER.debug("Azure AD App Token was acquired.")
+
+    # Get the external IP address of the Home Assistant instance.
+    aiotimeout = aiohttp.ClientTimeout(total=timeout)
+
+    async with aiohttp.ClientSession(timeout=aiotimeout) as aiosession:
+        async with aiosession.get('https://api.ipify.org') as aioresp:
+            ipv4address = (await aioresp.text())
+
+    if not ipv4address:
+        _LOGGER.error("Failed to get external IP address from ipify API")
+    else:
+        _LOGGER.debug("External IP address is: %s", ipv4address)
+
+    # Create a PATCH request with the Azure Resource Manager REST API.
+    headers = {
+        "Authorization": "Bearer " + access_token,
```

Comment:
Use string formatting.

Reply:
Not sure if I understand your comment here. access_token is not a string.